cep-abi3.md

```diff
@@ -0,0 +1,235 @@
+
+&nbsp;&nbsp;<strong>B3</strong>:
+  `.pyc` files created are recorded in `<PREFIX>/conda-meta/<pkg>.json`
+  so that they are removed properly when the package is uninstalled.
+
+&nbsp;&nbsp;<strong>B4</strong>:
+  Entry points in `info/link.json` are materialised.
+
+### info/link.json file
+An example `info/link.json` for `noarch: python` looks like
+
+```json
+{
+  "noarch": {
+    "entry_points": [
+      "isympy = isympy:main"
+    ],
+    "type": "python"
+  },
+  "package_metadata_version": 1,
+  "preferred_env": "foo"
+}
+```
+
+An example for `info/link.json` for `noarch: generic` looks like
+```
+{
+  "noarch": {
+    "type": "generic"
+  },
+  "package_metadata_version": 1
+}
+```
+
+
+Here `preferred_env` is ignored by conda since 2017 and is not supported by
+other conda install tools. Therefore `info/link.json` is used exclusively
+for `noarch` packages and out of the two types, `noarch: generic` packages
+does not require any special action.
+
+### info/index.json file
+
+An example for a `noarch: python` recipe.
+
+```json
+{
+  "arch": null,
+  "build": "pyh2585a3b_103",
+  "build_number": 103,
+  "depends": [
+    "mpmath >=0.19",
+    "python >=3.8"
+  ],
+  "license": "BSD-3-Clause",
+  "license_family": "BSD",
+  "name": "sympy",
+  "noarch": "python",
+  "platform": null,
+  "subdir": "noarch",
+  "timestamp": 1718625708903,
+  "version": "1.12.1"
+}
+```
+
+### Current behaviour in solver tools
+
+Conda package upload tools like `anaconda-client` use `A1` to upload
+the package to the `noarch` subdir.
+
+Conda install tools have slightly different behaviour.
+
+Conda:
+1. Actions `B1, B2, B3` are applied for packages with `A3`.
+2. Action `B4` is applied for packages with `A4`.
+
+Micromamba:
+1. Actions `B1, B2, B3` are applied for packages with both `A2, A3`.
+2. Action `B4` is applied for packages with both `A2, A4`.
+
+
+## Implementation for abi3 packages
+
+### Support in install tools
+
+We require the following attributes in abi3 packages:
+
+&nbsp;&nbsp;<strong>C1</strong>:
+  They have `subdir: <platform>` where `<platform>` is the subdir
+  that the package was built for.
+
+&nbsp;&nbsp;<strong>C2</strong>:
+  They have `noarch: python`.
+
+&nbsp;&nbsp;<strong>C3</strong>:
+  `A2, A3, A4` are applied.
+
+This is compatible with `conda/mamba/micromamba` install tools
+currently.
+
+### Support in build tools
+
+This requires support from build tools to set `subdir: <platform>`.
+
+In particular recipe authors would set
+```
+build:
+  python_version_independent: true
+
+requirements:
+  host:
+    - python
+    - python-abi3
+```
+which would make the build tool
+
+&nbsp;&nbsp;<strong>D1</strong>:
+Set `noarch: python` in `info/index.json`.
+
+Note that `python-abi3` would set the runtime requirements.
+This is explicitly required from recipe authors so that we do not
+restrict this CEP to ABI3 packages and allow the possibility for ABI4 etc.
+
+An example `python-abi3=3.8` package would set itself in its
```

Comment:
Minor: perhaps this should use the same `{{ python_min }}` syntax as was recently introduced for noarch packages?

Reply:
Unlike python which needs multiple values for arch builds and a single value for noarch builds, we only need a single value for abi3 builds. So we can set
```yaml
python_abi3:
  - 3.8
```
in our global pinning file in conda-forge and anaconda.

Reply:
Okay, that makes sense. Some of the "automatically bump" rationale of `python_min` probably still applies, but this dependency is indeed easier to handle than `python`.

Reply:
> we only need a single value for abi3 builds.

That's not guaranteed AFAIU. Despite the ABI being stable, new features get added in newer python versions, and if anything depends on those features, you might need both the oldest and a newer abi3 build.

For example, cryptography [builds](https://github.com/pyca/cryptography/blob/644dcafecf47dfd598302b35dbd53c6af3189fca/.github/workflows/wheel-builder.yml#L68-L69) for both the py37 & py39 limited API. In the future, it's quite likely that other features (e.g. nogil) will similarly require a newer baseline for some features.

Of course, individual feedstocks will be able to override `python_abi3` where necessary (and we may get away to only have a single value in the global pinning), but conceptually, I don't think we can assume it'll be one value.

Reply:
Sure, but what I want to emphasize is that we don't need different requirements based on `arch` vs `noarch`.